lib/acme/client/resources/challenges/base.rb

```diff
@@ -1,39 +1,43 @@
+# frozen_string_literal: true
+
 class Acme::Client::Resources::Challenges::Base
-  attr_reader :authorization, :status, :uri, :token, :error
+  attr_reader :status, :url, :token, :error
 
-  def initialize(authorization)
-    @authorization = authorization
+  def initialize(client, **arguments)
+    @client = client
+    assign_attributes(arguments)
   end
 
-  def client
-    authorization.client
+  def challenge_type
+    self.class::CHALLENGE_TYPE
```

Comment:
Where is `CHALLENGE_TYPE` set? I don't see `Acme::Client::Resources::Challenges::Base::CHALLENGE_TYPE` defined…

Reply:
`Base` is an abstract class ,`CHALLENGE_TYPE` is expected to be defined in the child classes. This is also why I do the `self.class::` dance because otherwise it would attempt to find it in the scope of `Base`.

Reply:
Ah, that totally makes sense.